src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -381,115 +353,131 @@ public override int Read(byte[] array, int offset, int count)
 
         internal unsafe int Read(byte[] array, int offset, int count, int timeout)
         {
-            if (_handle == null) throw new ObjectDisposedException(SR.Port_not_open);
-            if (array == null)
-                throw new ArgumentNullException(nameof(array));
-            if (offset < 0)
-                throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedNonNegNumRequired);
-            if (count < 0)
-                throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedNonNegNumRequired);
-            if (array.Length - offset < count)
-                throw new ArgumentException(SR.Argument_InvalidOffLen);
-            if (count == 0) return 0; // return immediately if no bytes requested; no need for overhead.
-
-            if (timeout != 0)
+            using (CancellationTokenSource cts = GetCancellationTokenSourceFromTimeout(timeout))
             {
-                Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
-                Interop.Sys.Poll(_handle, Interop.Sys.PollEvents.POLLIN | Interop.Sys.PollEvents.POLLERR, timeout,out events);
+                Task<int> t = ReadAsync(array, offset, count, cts?.Token ?? CancellationToken.None);
 
-                if ((events & (Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLNVAL)) != 0)
+                try
                 {
-                    throw new IOException();
+                    return t.GetAwaiter().GetResult();
                 }
-
-                if ( (events & Interop.Sys.PollEvents.POLLIN) == 0)
+                catch (OperationCanceledException)
                 {
                     throw new TimeoutException();
                 }
             }
+        }
 
-            int numBytes;
-            fixed (byte* bufPtr = array)
-            {
-                numBytes = Interop.Sys.Read(_handle, bufPtr + offset, count);
-            }
+        public override int EndRead(IAsyncResult asyncResult)
+            => EndReadWrite(asyncResult);
 
-            if (numBytes < 0)
-            {
-                if (Interop.Error.EWOULDBLOCK == Interop.Sys.GetLastError())
-                {
-                    throw new TimeoutException();
-                }
-                throw new IOException();
-            }
+        public override Task<int> ReadAsync(byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            CheckReadWriteArguments(array, offset, count);
 
-            return numBytes;
+            if (count == 0)
+                return Task<int>.FromResult(0); // return immediately if no bytes requested; no need for overhead.
+
+            Memory<byte> buffer = new Memory<byte>(array, offset, count);
```

Comment:
>Memory<byte> bu [](start = 12, length = 15)

would be any concern allocating Memory object with every ReadASync call?

Reply:
In general I do not expect anything to be slower than the reading/writing itself (the max speed for serial port is super slow comparing to anything else) - since it's a struct I think the overhead should be tiny

Reply:
`Memory<T>` is a struct.